### ISSUE TITLE
Lots of ASN1Errors when monitoring WoSign's CT Log using logobserver.py

### ISSUE THREAD

**Issue by linshunghuang** [google/certificate-transparency#1078]
Running python/utilities/logobserver/logobserver.py against https://ct.wosign.com raises a lot of ASN1Errors (and therefore excluded from scan results), for instance:
```
ASN1Error: Corrupt time: 151207195316
```
These errors started to appear since the certificate entry 11309. Looking at the certificate entry 11309 (https://crt.sh/?id=11142769) of https://ct.wosign.com, the `not_before` timestamp looks reasonable, so I'm wondering if there's a bug in the python ASN1 parser, or are these certificates indeed malformed? Thanks!

Repro steps:
1. Add WoSign CT Log to python/ct/config/logs.config
```
ctlog:{
  log_server:'https://ct.wosign.com'
  log_id:'nk/3PcPOIgtpIXyJnkaAdqv414Y21cz8haMadWKLqIs='
  public_key_info:{
    type:ECDSA
    pem_key:'-----BEGIN ECDSA PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0D'
    'AQcDQgAE1+wvK3VPN7yjQ7qLZWY8fWrlDCqmwuUm/gx9TnzwOrzi0yLcAdAfbkOcXG6DrZ'
    'wV9sSNYLUdu6NiaX7rp6oBmw==\n-----END ECDSA PUBLIC KEY-----'
  }
}
```
2.
```
certificate-transparency$ cd python/
python$ export PYTHONPATH=${PYTHONPATH}:$(pwd)
python$ python utilities/logobserver/logobserver.py
```
Wait a minute or so and observe:
```
WARNING:root:Monitor state file not found, assuming first run.
ERROR:root:Error scanning certificate 11309 in batch <11000, 11430> - it will be excluded from the scan results
Traceback (most recent call last):
  File "/MY_PATH/certificate-transparency/python/ct/client/reporter.py", line 52, in _scan_der_cert
    partial_result += check.check(certificate) or []
  File "/MY_PATH/certificate-transparency/python/ct/cert_analysis/validity.py", line 47, in check
    not_before = certificate.not_before()
  File "/MY_PATH/certificate-transparency/python/ct/crypto/cert.py", line 543, in not_before
    return (self._asn1_cert["tbsCertificate"]["validity"]["notBefore"].
  File "/MY_PATH/certificate-transparency/python/ct/crypto/asn1/x509_time.py", line 33, in gmtime
    raise error.ASN1Error("Corrupt time: %s" % self._value)
ASN1Error: Corrupt time: 151207195316
ERROR:root:Error scanning certificate 11310 in batch <11000, 11430> - it will be excluded from the scan results
Traceback (most recent call last):
  File "/MY_PATH/certificate-transparency/python/ct/client/reporter.py", line 52, in _scan_der_cert
    partial_result += check.check(certificate) or []
  File "/MY_PATH/certificate-transparency/python/ct/cert_analysis/validity.py", line 47, in check
    not_before = certificate.not_before()
  File "/MY_PATH/certificate-transparency/python/ct/crypto/cert.py", line 543, in not_before
    return (self._asn1_cert["tbsCertificate"]["validity"]["notBefore"].
  File "/MY_PATH/certificate-transparency/python/ct/crypto/asn1/x509_time.py", line 33, in gmtime
    raise error.ASN1Error("Corrupt time: %s" % self._value)
ASN1Error: Corrupt time: 151207195349
ERROR:root:Error scanning certificate 11311 in batch <11000, 11430> - it will be excluded from the scan results
Traceback (most recent call last):
  File "/MY_PATH/certificate-transparency/python/ct/client/reporter.py", line 52, in _scan_der_cert
    partial_result += check.check(certificate) or []
  File "/MY_PATH/certificate-transparency/python/ct/cert_analysis/validity.py", line 47, in check
    not_before = certificate.not_before()
  File "/MY_PATH/certificate-transparency/python/ct/crypto/cert.py", line 543, in not_before
    return (self._asn1_cert["tbsCertificate"]["validity"]["notBefore"].
  File "/MY_PATH/certificate-transparency/python/ct/crypto/asn1/x509_time.py", line 33, in gmtime
    raise error.ASN1Error("Corrupt time: %s" % self._value)
ASN1Error: Corrupt time: 151207195357
ERROR:root:Error scanning certificate 11313 in batch <11000, 11430> - it will be excluded from the scan results
Traceback (most recent call last):
  File "/MY_PATH/certificate-transparency/python/ct/client/reporter.py", line 52, in _scan_der_cert
    partial_result += check.check(certificate) or []
  File "/MY_PATH/certificate-transparency/python/ct/cert_analysis/validity.py", line 47, in check
    not_before = certificate.not_before()
  File "/MY_PATH/certificate-transparency/python/ct/crypto/cert.py", line 543, in not_before
    return (self._asn1_cert["tbsCertificate"]["validity"]["notBefore"].
  File "/MY_PATH/certificate-transparency/python/ct/crypto/asn1/x509_time.py", line 33, in gmtime
    raise error.ASN1Error("Corrupt time: %s" % self._value)
ASN1Error: Corrupt time: 151208091403
ERROR:root:Error scanning certificate 11314 in batch <11000, 11430> - it will be excluded from the scan results
Traceback (most recent call last):
  File "/MY_PATH/certificate-transparency/python/ct/client/reporter.py", line 52, in _scan_der_cert
    partial_result += check.check(certificate) or []
  File "/MY_PATH/certificate-transparency/python/ct/cert_analysis/validity.py", line 47, in check
    not_before = certificate.not_before()
  File "/MY_PATH/certificate-transparency/python/ct/crypto/cert.py", line 543, in not_before
    return (self._asn1_cert["tbsCertificate"]["validity"]["notBefore"].
  File "/MY_PATH/certificate-transparency/python/ct/crypto/asn1/x509_time.py", line 33, in gmtime
    raise error.ASN1Error("Corrupt time: %s" % self._value)
ASN1Error: Corrupt time: 151208124859

(a lot more ASN1Errors omitted...)
```

**Comment by benlaurie** [comment]
On Thu, 10 Dec 2015 at 07:44 David Huang <notifications@github.com> wrote:

> Running python/utilities/logobserver/logobserver.py against
> https://ct.wosign.com raises a lot of ASN1Errors (and therefore excluded
> from scan results), for instance:
>
> ASN1Error: Corrupt time: 151207195316
>
> These errors started to appear since the certificate entry 11309. Looking
> at the certificate entry 11309 (https://crt.sh/?id=11142769) of
> https://ct.wosign.com, the not_before timestamp looks reasonable, so I'm
> wondering if there's a bug in the python ASN1 parser, or are these
> certificates indeed malformed? Thanks!
>

The certificates are malformed - if you look at the timestamps, they look
like, e.g. "151207195316" - this is not a legal format for UTCTIME: it
should end with a Z or with a timezone offset.

**Comment by linshunghuang** [comment]
Thanks! I was curious is because crt.sh and the openssl x509 command line seemed to be able to parse the certificate and show a reasonable notBefore timestamp without errors.

The other issue is that these certificates are missing in the cert_db. Should these malformed certificates still be stored regardless of ASN1Errors?

**Comment by benlaurie** [comment]
On Thu, 10 Dec 2015 at 17:58 David Huang <notifications@github.com> wrote:

> Thanks! I was curious is because crt.sh and the openssl x509 command line
> seemed to be able to parse the certificate and show a reasonable notBefore
> timestamp without errors.
>

I have not checked, but I assume it is allowing them to not strictly comply
with the standards. This could be regarded as a bug.


> The other issue is that these certificates are missing in the cert_db.
> Should these malformed certificates still be stored regardless of
> ASN1Errors?
>

No strong opinion - if you can't parse them, they're not much use to you.
In general, the approach we've taken is to add "non-strict" parsing to the
ASN1 code. This particular error has not been allowed for, but could be
somwhere around here:

https://github.com/google/certificate-transparency/blob/879395cc4aeecdbe0452249b59c84795bb231638/python/ct/crypto/asn1/x509_time.py#L48

**Comment by liangdong2718** [comment]
Actually, this malformed certificates is generate by openssl, this is why crt.sh and openssl can parse it.  I use openssl ca function and forget to add an "Z" in startdate parameter.
And I'm so sorry for this malformed certificates, so I simple write a patch to fix it, if you don't like to see the exception raise, you can patch the attach file. Thanks!
[python_utctime_patch.txt](https://github.com/google/certificate-transparency/files/65043/python_utctime_patch.txt)

**Comment by benlaurie** [comment]
On Thu, 17 Dec 2015 at 07:25 dliang <notifications@github.com> wrote:

> Actually, this malformed certificates is generate by openssl, this is why
> crt.sh and openssl can parse it. I use openssl ca function and forget to
> add an "Z" in startdate parameter.
>

This is a bug in OpenSSL - I have a fix under review.

**Comment by linshunghuang** [comment]
Fixed by #1110
